Python Library for publishing events and fetching messages from a queue

noeliaschinder opened the pull request:
Created a new library for publishing events and reading messages from a queue.

For now, it will only contain adapters for SNS events and SQS queues. In the future we could add more if we decide to.

Readme file content:

# clever-events-python-library

The `clever-events-python-library` is a Python library designed to facilitate the publishing and management of events and messages using AWS services such as SNS (Simple Notification Service) and SQS (Simple Queue Service). This library provides a simple and efficient way to publish events both synchronously and asynchronously, as well as to fetch and delete messages from queues. It abstracts the complexities of interacting with AWS services, allowing developers to focus on building their applications without worrying about the underlying infrastructure.

## Usage Examples

### SNSAdapter with EventPublisher

#### Synchronous Publishing

```python
from clever_events_library.events.adapters import SNSAdapter
from clever_events_library.events.event_publisher import EventPublisher

sns_adapter = SNSAdapter({
    'aws_region': 'us-east-1',
    'aws_key': 'AWS_KEY',
    'aws_secret': 'AWS_SECRET',
    'aws_account_id': 'AWS_ACCOUNT_ID',
})

EventPublisher(event_adapter=sns_adapter).sync_publish(
    event_name='test_sns_topic',
    message_data={
        'message': {'test': 'TEST message'},
        'message_attributes': {
            'test1': 'test attributes 1',
        }
    }
)
```

#### Asynchronous Publishing

```python
from clever_events_library.events.adapters import SNSAdapter
from clever_events_library.events.event_publisher import EventPublisher
import asyncio

sns_adapter = SNSAdapter({
    'aws_region': 'us-east-1',
    'aws_key': 'AWS_KEY',
    'aws_secret': 'AWS_SECRET',
    'aws_account_id': 'AWS_ACCOUNT_ID',
})

loop = asyncio.new_event_loop()
asyncio.set_event_loop(loop)

event_publisher = EventPublisher(event_adapter=sns_adapter)
loop.run_until_complete(event_publisher.async_publish(
    event_name='test_sns_topic',
    message_data={
        'message': {'test': "TEST message"},
        'message_attributes': {
            'test': 'test attributes',
        }
    }
))
```

### SQSAdapter with QueueManager

#### Fetching and Deleting Messages

```python
from clever_events_library.queues.adapters import SQSAdapter
from clever_events_library.queues.queue_manager import QueueManager

sqs_adapter = SQSAdapter(
    {
        'aws_region': 'us-east-1',
        'aws_key': 'AWS_KEY',
        'aws_secret': 'AWS_SECRET',
        'aws_account_id': 'AWS_ACCOUNT_ID',
    }
)

sqs_adapter.set_await_time(20)

queue_manager = QueueManager(queue_adapter=sqs_adapter)
messages = queue_manager.fetch_messages(queue_name="noelias_test_queue", max_number_of_messages=10)

for message in messages:
    queue_manager.delete_message(queue_name="noelias_test_queue", message_id=message['message_receipt_handle'])
```

### AWS variables set up

AWS variables like region, account id and credentials can be configured as shown in examples above or can be set up in an environment (.env) file as follows:

```
AWS_REGION=AWS_REGION
AWS_ACCESS_KEY_ID=AWS_KEY
AWS_SECRET_ACCESS_KEY=AWS_SECRET
AWS_ACCOUNT_ID=AWS_ACCOUNT_ID
```

When those variables are added to .env file, it's not needed to explicitly include them in the adapter initialization.

So, the following code:

```
sqs_adapter = SQSAdapter(
    {
        'aws_region': 'us-east-1',
        'aws_key': 'AWS_KEY',
        'aws_secret': 'AWS_SECRET',
        'aws_account_id': 'AWS_ACCOUNT_ID',
    }
)
```

can be replaced with:

```
sqs_adapter = SQSAdapter()
```

Same applies to SNSAdapter.

nickcluc commented on README.md:
```diff
@@ -1 +1,49 @@
+messages = queue_manager.fetch_messages(queue_name="noelias_test_queue", max_number_of_messages=10)
+
+for message in messages:
+    queue_manager.delete_message(queue_name="noelias_test_queue", message_id=message['message_receipt_handle'])
```
nice so the idea would be that we pull a message or group of messages off, then do some work with it, then delete it?

Do you imagine there is there a process that manages pulling these off the queue? Not pertient for any changes here, but I'm excited to see how this would be implemented!

Irapoam replied:
Reading sqs messages requires you to “read the message successfully”
In the queue settings there is a limit on the number of attempts to read from the queue. If the limit is reached, the message goes to the dead queue.

noeliaschinder replied:
Hey @nickcluc , you mean if there's a use case where we would fetch the messages and delete them right away? Without having to explicitly call the delete message method? I thought about doing that before but I think it's a bit risky. The consumer will process the messages after deleting them but what happens if something fails during that processing? The messages won't be in the queue anymore for a 2nd try, so unless the app saves the messages somewhere, they will be lost. That's how I see it, I might be wrong . What do you think?

nickcluc replied:
@noeliaschinder I think thats a good question. To me, the purpose of the library should be a standardized way to _get_ the messages to process, but once the library hands it off to the service using it, the library should consider their part in the transaction done. errors with processing should ultimately be handled by the processor itself, as long as there wasn't an error in the action of posting or getting the message itself.

I was just curious, mostly, about the api for sqs. And so m aybe. iframed my question wrong, but does the act of pulling the message from the queue not also remove/delete it from the queue? I was just surprised we had to explicitly delete it also.

noeliaschinder replied:
Hey @nickcluc , nope, the act of pulling a message from the queue does not remove it from the queue, we need to explicitly do it.

nickcluc commented on clever_events_library/events/adapters/sns_adapter.py:
```diff
@@ -0,0 +1,57 @@
+
+    def publish(self, event_name, message_data):
+        if "message" not in message_data:
+            raise ValueError("message is required in message_data")
```
Is this going to raise if the string "message" is missing?

noeliaschinder replied:
Hey @nickcluc , this is a dict but now that you mention it, I will add some changes in the code to make it clear the type of each variable.

nickcluc commented on clever_events_library/events/adapters/sns_adapter.py:
```diff
@@ -0,0 +1,57 @@
+        )
+
+    def _get_topic_arn(self, topic_name):
+        return "arn:aws:sns:{}:{}:{}".format(self.aws_region, self.aws_account_id, topic_name)
```
🎱  do you think we should let the client supply the entire arn or is the topic name enough? 

My only thought here is if we somehow get the arn building wrong or something with how the arn format changes in SQS (which is probably unlikely), it woudl require an update to the library vs just updating the config in the app.

Tradeoff being that we would need to update the arn in _every_ app, opposed to one place in the library.

(🎱  means i don't have a strong opinion either way, just wanted to put the question out there)

noeliaschinder replied:
Hey @nickcluc , I vote that we just send the queue name, as you've said, if AWS changes the way they structure ARNs, we'll only have to update the library and not all the consumer apps. Also, I prefer having logic like this in the library so apps don't have to care about it.

nickcluc commented on clever_events_library/queues/adapters/sqs_adapter.py:
```diff
@@ -0,0 +1,77 @@
+        required_keys = ["aws_region", "aws_key", "aws_secret", "aws_account_id"]
+        for key in required_keys:
+            if key not in client_config:
+                raise ValueError(f"{key} is required in client_config")
```
👍

nickcluc commented on clever_events_library/queues/adapters/sqs_adapter.py:
```diff
@@ -0,0 +1,77 @@
+        )
+
+        messages = []
+        if "Messages" in response:
```
Kind of the same question as above, ut does this block require us to have the string "Messages" in the response? Or is this a dict key?

noeliaschinder replied:
It is a dict key. In python, this is a way of checking if a key is in a dictionary.

nickcluc commented on README.md:
```diff
@@ -1 +1,49 @@
+from clever_events_library.events.adapters import SNSAdapter
+from clever_events_library.events.event_publisher import EventPublisher
+
+sns_adapter = SNSAdapter({
+    'aws_region': 'us-east-1',
+    'aws_key': 'AWS_KEY',
+    'aws_secret': 'AWS_SECRET',
+    'aws_account_id': 'AWS_ACCOUNT_ID',
+})
```
Are we going to need to define this at runtime in the app or does django have an initialization system where this can be defined once, then `sns_adapter` is available wherever we want to use it?

noeliaschinder replied:
That's a good question. IDK the answer but I'll investigate.

nickcluc commented on README.md:
```diff
@@ -1 +1,49 @@
+from clever_events_library.queues.queue_manager import QueueManager
+
+
+sqs_adapter = SQSAdapter(
+    {
+        'aws_region': 'us-east-1',
+        'aws_key': 'AWS_KEY',
+        'aws_secret': 'AWS_SECRET',
+        'aws_account_id': 'AWS_ACCOUNT_ID',
+        'await_time': 20
+    }
+)
```
Same Q as above

nickcluc commented:
> If it's something new, it might make sense to start with asynchronous support. Instead of using boto3 we could use aioboto3

@Irapoam can the async part be implemented in the application? The idea would be to let the client using the library choose how they want to implement (although you're right, it probably good practice to implement it as async)

Irapoam commented:
> > If it's something new, it might make sense to start with asynchronous support. Instead of using boto3 we could use aioboto3
> 
> @Irapoam can the async part be implemented in the application? The idea would be to let the client using the library choose how they want to implement (although you're right, it's probably good practice to implement it as async)

Yes, the async part can definitely be handled at the application level. However, if we provide built-in async support, we give users the flexibility to use it directly without extra effort. Since aioboto3 is a drop-in replacement for boto3, we could design the library to support both sync and async calls, allowing the client to choose based on their needs.

Irapoam commented:
I was a little late to suggest this, but during my presentation I used a library called [pyinsole](https://github.com/pyinsole/pyinsole/tree/main)

For the queue reading part, it already has good asynchronous support with read confirmation, error handling, etc.
So maybe we can take something good from there

Irapoam commented on setup.cfg:
```diff
@@ -0,0 +1,7 @@
+install_requires =
+    boto3
+    botocore
+python_requires = >=3.7
```
I think we need to go straight to python 3.12+

noeliaschinder replied:
hmm if we do that right now, this library won't be used by Messaging Service. I think it uses 3.9 or so.

Irapoam replied:
let's update the python version in messaging service.  it will be simple to update there

noeliaschinder replied:
ok cool

Irapoam commented on clever_events_library/queues/adapters/sqs_adapter.py:
```diff
@@ -0,0 +1,77 @@
+                    else {}
+                )
+                message_data = json.loads(message_body["Message"])
+                messages.append(
```
Can we use generators instead of a normal list?

yield 
```suggestion
                message =  {
                        "message_id": message_id,
                        "message_receipt_handle": message_receipt_handle,
                        "message_data": message_data,
                        "message_attributes": message_attributes,
                    }
                yield message
```

noeliaschinder replied:
Good callout ! Done.

Irapoam commented:
for a library, I think it's nice to use typehint and docstrings in the methods.

Having the definition in the functions, we can take advantage of this python benefit and type an exclamation mark next to the function name in case we want to know the function definitions
![Screenshot from 2025-02-19 17-19-06](https://github.com/user-attachments/assets/f8f88cf3-6586-46a2-ae95-4432597e0f45)

noeliaschinder commented:
@nickcluc and @Irapoam , I've pushed lot of changes to the library, would you take a look again on Monday:

Changes recap:
* Added type hints and docstrings to public methods (must be all but let me know if you find one that has missing docs) (I <3 copilot).
* Added **aiboto3** so we can do async calls to publish messages to SNS topics .
* Updated the code so AWS variables can be read from a env variables instead of having to include them in the adapters initizializations.
* Added a mixin for AWS Adapters because there is some shared code that was repeated in both SNS and SQS adapters.
* Updated required python version to 3.12
* Updated tests accordingly (ofc with the huge help of copilot here too)

@nickcluc I still owe you an answer [here](https://github.com/clever-real-estate/clever-events-python-library/pull/1#discussion_r1962276934). I'll try implementing the library with latest changes into AS on Monday and will let you know. 

CC @rymccue and @jlien just in case you want to review this too (on Monday ofc).

noeliaschinder commented:
> I was a little late to suggest this, but during my presentation I used a library called [pyinsole](https://github.com/pyinsole/pyinsole/tree/main)
> 
> For the queue reading part, it already has good asynchronous support with read confirmation, error handling, etc. So maybe we can take something good from there

Hey @Irapoam , which feature of this library do you think it will be useful for us? I can't see we'll need anything that is not implemented in our library but I might be missing something.

nickcluc reviewed (approved):
Nice! LGTM! LMK if you need help cleaning up the commits and lets get this merged!
